compiler/src/typed/typecore.re

```diff
@@ -2690,29 +2756,96 @@ let report_error = (env, ppf) =>
     )
   | Apply_non_function(typ) => {
       reset_and_mark_loops(typ);
-      switch (repr(typ).desc) {
-      | TTyArrow(_) =>
+      fprintf(
+        ppf,
+        "@[<v>@[<2>This expression has type@ %a@]@ %s@]",
+        type_expr,
+        typ,
+        "This is not a function; it cannot be applied.",
+      );
+    }
+  | Apply_too_many_arguments(typ, unused_tyargs) => {
+      reset_and_mark_loops(typ);
+      fprintf(ppf, "@[<v>@[<2>This function has type@ %a@]", type_expr, typ);
+      fprintf(ppf, "@ @[It is applied to too many arguments.@]@]");
+      let unused_optional_arguments =
+        List.filter_map(
+          ((l, _)) =>
+            switch (l) {
+            | Default({txt: name}) => Some(name)
+            | _ => None
+            },
+          unused_tyargs,
+        );
+      let oxford = (ppf, args) => {
+        let rec oxford = (ppf, args) =>
+          switch (args) {
+          | [] => ()
+          | [arg1, arg2] => fprintf(ppf, "%s, or %s", arg1, arg2)
+          | [arg1, ...args] =>
+            fprintf(ppf, "%s, ", arg1);
+            oxford(ppf, args);
+          };
+
+        switch (args) {
+        | [] => ()
+        | [arg] => fprintf(ppf, "%s", arg)
+        | [arg1, arg2] => fprintf(ppf, "%s or %s", arg1, arg2)
+        | _ => oxford(ppf, args)
+        };
+      };
+      switch (unused_optional_arguments) {
+      | [] => ()
+      | labels =>
         fprintf(
           ppf,
-          "@[<v>@[<2>This function has type@ %a@]",
-          type_expr,
-          typ,
-        );
+          "@ @[Did you mean to supply an optional argument with label %a?@]@]",
```

Comment:
```suggestion
          "@ @[Did you mean to supply an argument with label %a?@]@]",
```
I don't think this is correct but I'm trying to not use "optional" terminology, just "default" and "labeled", if we can.